File: backend/src/main/java/com/github/denvned/xodus/compare/backend/graphql/EntityType.java
package com.github.denvned.xodus.compare.backend.graphql;

import com.github.denvned.graphql.annotations.GraphQLField;
import com.github.denvned.graphql.annotations.GraphQLName;
import com.github.denvned.graphql.annotations.GraphQLNonNull;
import com.github.denvned.xodus.compare.ComparisonStoreNames;
import com.github.denvned.xodus.compare.backend.ComparisonStoreProvider;
import jetbrains.exodus.entitystore.Entity;
import jetbrains.exodus.entitystore.EntityIterable;
import jetbrains.exodus.entitystore.StoreTransaction;

public final class EntityType extends AbstractEntityBasedNode {
    public EntityType(Entity entity) {
        super(entity);
    }

    @GraphQLField @GraphQLNonNull
    public long getLocalId() {
        return entity.getId().getLocalId();
    }

    @GraphQLField @GraphQLNonNull
    public Comparison getComparison() {
        return new Comparison(entity.getLink(ComparisonStoreNames.EntityType.COMPARISON));
    }

    @GraphQLField @GraphQLNonNull
    public int getTypeId() {
        return (int)entity.getProperty(ComparisonStoreNames.EntityType.ID);
    }

    @GraphQLField
    public String getOldName() {
        return (String)entity.getProperty(ComparisonStoreNames.EntityType.OLD_NAME);
    }

    @GraphQLField
    public String getNewName() {
        return (String)entity.getProperty(ComparisonStoreNames.EntityType.NEW_NAME);
    }

    @GraphQLField @GraphQLNonNull
    public AddedEntityConnection getAddedEntities(
        @GraphQLName("first") Integer first,
        @GraphQLName("after") String after
    ) {
        return new AddedEntityConnection(getEntities(
            ComparisonStoreNames.ADDED_ENTITY),
            first,
            after != null ? Long.parseLong(after) : null
        );
    }

    @GraphQLField @GraphQLNonNull
    public ChangedEntityConnection getChangedEntities(
        @GraphQLName("first") Integer first,
        @GraphQLName("after") String after
    ) {
        return new ChangedEntityConnection(getEntities(
            ComparisonStoreNames.CHANGED_ENTITY),
            first,
            after != null ? Long.parseLong(after) : null
        );
    }

    @GraphQLField @GraphQLNonNull
    public DeletedEntityConnection getDeletedEntities(
        @GraphQLName("first") Integer first,
        @GraphQLName("after") String after
    ) {
        return new DeletedEntityConnection(getEntities(
            ComparisonStoreNames.DELETED_ENTITY),
            first,
            after != null ? Long.parseLong(after) : null
        );
    }

    private EntityIterable getEntities(String changeType) {
        StoreTransaction txn = ComparisonStoreProvider.getStore().getCurrentTransaction();

        return txn.findLinks(
            changeType,
            entity,
            ComparisonStoreNames.Entity.TYPE
        );
    }
}

File: backend/src/main/kotlin/com/github/denvned/xodus/compare/backend/graphql/EntityType.kt
package com.github.denvned.xodus.compare.backend.graphql

import com.github.denvned.xodus.compare.ComparisonStoreNames
import com.github.denvned.xodus.compare.backend.ComparisonStoreProvider
import jetbrains.exodus.entitystore.Entity
import jetbrains.exodus.entitystore.EntityIterable

class EntityType(entity: Entity) : AbstractEntityBasedNode(entity) {

  val localId get() = entity.id.localId

  val comparison get() = Comparison(entity.getLink(ComparisonStoreNames.EntityType.COMPARISON)!!)

  val typeId get() = entity.getProperty(ComparisonStoreNames.EntityType.ID) as Int

  val oldName get() = entity.getProperty(ComparisonStoreNames.EntityType.OLD_NAME) as String?

  val newName get() = entity.getProperty(ComparisonStoreNames.EntityType.NEW_NAME) as String?

  fun getAddedEntities(first: Int?, after: String?) =
      AddedEntityConnection(getEntities(ComparisonStoreNames.ADDED_ENTITY), first, after?.toLong())

  fun getChangedEntities(first: Int?, after: String?) =
    ChangedEntityConnection(getEntities(ComparisonStoreNames.CHANGED_ENTITY), first, after?.toLong())

  fun getDeletedEntities(first: Int?, after: String?) =
    DeletedEntityConnection(getEntities(ComparisonStoreNames.DELETED_ENTITY), first, after?.toLong())

  private fun getEntities(changeType: String): EntityIterable {
    val txn = ComparisonStoreProvider.store.currentTransaction!!
    return txn.findLinks(changeType, entity, ComparisonStoreNames.Entity.TYPE)
  }
}
